### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by romanfurman6**

### CODE REVIEW

File: 5-tests.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+// Usage in tests:
+let testScheduler = TestScheduler(initialClock: 0)
+let date = Date()
+let dateProducer = { return date }
+let viewModel = ViewModel(authService: testAuthService, scheduler: testScheduler, dateProducer: dateProducer)
+```
+
+## UI Tests
+
+UI tests help a lot with testing View Controllers and coordination, but take much more time to write and run than unit tests. Consider writing UI tests for long-term projects, projects with complicated UI design and navigation.
```
Comment (EvgenyMatviyenko) [inline 5-tests.md]: I suggest to remove this section at all or investigate deeper and add more info, cause current state of it doesn't help to understand the topic anyhow

Comment (romanfurman6) [reply 5-tests.md]: @annakorolyova do you want add something?

Comment (annakorolyova) [reply 5-tests.md]: I agree that this section needs to be more informative. However, I'd like to keep it there. I can continue working on it a bit later, when I'm back from vacation.
@EvgenyMatviyenko can you suggest any ways to make this more helpful? Providing examples of UI tests? Investigating when and how UI testing is most useful?

File: 5-tests.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+### General Rules
+
+##### 1. We test our code.
+
+Although we don't aim for 100% code coverage, we keep it at a reasonable level. Ultimately, you decide by yourself what part of a specific project it is crucial to cover with tests and what kind of tests you need.
+
+##### 2. Naming
+
+Test class should be named as `\(TestingClassName)Tests.swift`.
+
```
Comment (aLHF) [inline 5-tests.md]: Have you considered such naming: "test_behaviorBeingTested_contextItsTestedUnder"

File: 5-tests.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+*(WARNING: it may work wrong for types with implementation of `CustomDebugStringConvertible`)*
+
+```swift
+XCTAsserEqual(events.debugDescription, expectedEvents.debugDescription)
+```
+
+###### Why do we do this?
+Void is not equatable, and we can't compare it. In debug description we will get some information about out events (example: `next(()) @ 200`) where `next(())` our void event and `@ 200` it's time  in milliseconds when it comes.
+
```
Comment (aLHF) [inline 5-tests.md]: typo: "about out"

File: 5-tests.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    self.testFetchUser = testFetchUser
+  }
+}
+let testScheduler = TestScheduler(initialClock: 0)
+let date = Date()
+let currentDateFactory = { return date }
+let testUser = User(id: 1)
+
+// I recommend you mock only this `func` or `properties` which you will use in viewModel. 
+// We set a default value for testFetchUser to Observable.empty().
```
Comment (aLHF) [inline 5-tests.md]: we don't need "this" here, just you mock only `funcs` or `properties` which...

File: 5-tests.md (hunk 2 above)
Comment (arthur-here) [reply 5-tests.md]: Or that: `test_contextAndBehvaior_expectedResult`, e.g. `test_clickLoginButtonWithoutCredentials_EmitsError()`

Comment (romanfurman6) [reply 5-tests.md]: @aLHF & @arthur-here 
Ye i thinked about adding a third part of naming (test_`what we test`_`result`())
But on practice, it's not make any sense, every your test function name will contain `_expectedResult` :D

**Pull request merged**